### PR TITLE
Adding top level README for docs

### PR DESCRIPTION
**Pull request by rvs**
This is a long promised top level set of docs. Think of it as a table of contents of sorts. @deitch I will need your magic to make sure we can review, iterate and merge it relatively quickly. The idea here is not to be exhaustive, but rather provide enough of a skeleton on which to hang all the follow up meaty documentation.

For the ease of review: here's the direct link https://github.com/rvs/eve/blob/docs2/docs/README.md

### CODE REVIEW

**Comment by deitch** [comment]
You pulled all of the docs out of `pkg/pillar/docs/` and centralized them and made them clean. That alone makes this PR a "good thing". :-)

**Comment by deitch** [comment]
By the way, it needs a rebase.

File: docs/README.md [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+7. [Runtime Lifecycle](#runtime-lifecycle)
+8. [Building EVE](#building-eve)
+9. [EVE Internals](#eve-internals)
+
+## Introduction
+
+Edge Virtualization Engine (EVE) is a secure by design operating system
+that supports running *Edge Containers* on compute devices deployed in
+the field. These devices can be IoT gateways, Industrial PCs or general
+purpose ruggedized computers and we will refer to them as *Edge Nodes*.
```
Comment (deitch) [inline docs/README.md]: "or general purpose ruggedized computers; collectively, these are referred to as *Edge Nodes*."

Comment (deitch) [reply docs/README.md]: By the way, I like this easing into the terms starting from a place most technologists are comfortable without getting into a formal style. It works well.

File: docs/README.md [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+EVE is based on a type-1 hypervisor and it does not directly support
+traditional POSIX-like application and processes. All applications
+running on EVE have to be represented as [Edge Containers](ECOS.md) which
+can be either Virtual Machines, Containers or Unikernels. EVE presents a
```
Comment (deitch) [inline docs/README.md]: "either / or" by definition means 2 options. We have 3. So we have to make it, "can be one of: Virtual Machines, Containers or Unikernels."

File: docs/README.md [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+somewhat of a novel design of an operating systems, but at the same time
```
Comment (deitch) [inline docs/README.md]: "somewhat novel design" ("of a" is extraneous).

Comment (deitch) [reply docs/README.md]: Can we replace the "but at the same time it is" with simply, "somewhat novel design of an operating system, while heavily inspired by ideas from..." ?

File: docs/README.md [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+it is heavily inspired by ideas from [Qubes OS](https://www.qubes-os.org/),
+[ChromeOS](https://www.chromium.org/chromium-os/chromiumos-design-docs),
+[Core OS](https://coreos.com/) and [Smart OS](https://wiki.smartos.org/).
+
+To put it another way, EVE is all you need to put on your Edge Nodes in order
```
Comment (deitch) [inline docs/README.md]: Get rid of "To put it another way"; it weakens the statement. Just make it, "EVE is all you need on your Edge Nodes to run any kind of workload on the edge... Edge Containers."

File: docs/README.md [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+to run Edge Containers there. You don't need to install traditional
+(or embedded) operating systems, hypervisors, container engines and middle
```
Comment (deitch) [inline docs/README.md]: "or middleware" ("or" instead of "and", and "middleware" should be a single word).

File: docs/README.md [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+ware. EVE provides all of that in a tightly integrated, secure and space
+conscious package.
+
+EVE's promise to its users is simple:
+* Zero touch: you should never have physical access to your Edge Node,
```
Comment (deitch) [inline docs/README.md]: "you should never need physical access" - "never have" implies you cannot which restricts the adopter; "never need" implies "need not", which liberates them.

File: docs/README.md [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+It is expected to evolve significantly based on community feedback and constant
+addition of new functionality to EVE. The structure of this [README](README.md)
+is aimed at providing just enough information to vector you off to a relevant
+part  of the documentation. Even then, it can be too long for some readers and
+we  provide a companion [FAQ](FAQ.md) for those simply trying to look random
+facts  up. Finally, if you are interested in the gory details of how each
+subcomponent of EVE is implemented you should refer to respective subcomponents
+[docs folders](../pkg/) and especially [pillar docs](../pkg/pillar/docs/README.md).
+
+EVE leverages an enormous amount of open source projects and it would *not* have
```
Comment (deitch) [inline docs/README.md]: "leverages many open source projects" ("enormous" is a frightening word for corporate acquisition and legal departments).

File: docs/README.md [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+## Edge Containers
+
+Edge Container is a novel concept that allows an effective packaging and lifecycle
+management of the following types of workloads:
+
+* [Traditional Virtual Machines](https://en.wikipedia.org/wiki/Virtual_machine)
+* [Unikernels](http://unikernel.org/)
+* [Docker/OCI container](https://www.opencontainers.org)
+
+EVE is *not* a traditional operating system in that it doesn't support applications
```
Comment (deitch) [inline docs/README.md]: "doesn't support running workload applications" - it isn't that POSIX is disabled or it is something else, as much as we prevent it.

File: docs/README.md [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+represented as executable files and [POSIX processes](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_295)
+and expects all its application to be represented as Edge Containers.
+
+Think of it this way: Android has [APKs](https://en.wikipedia.org/wiki/Android_application_package),
```
Comment (deitch) [inline docs/README.md]: Eliminate "think of it this way". It works conversationally, but weakens the phrase in writing.

File: docs/README.md [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+[Edge Computing scenarious](https://www.zdnet.com/article/10-scenarios-where-edge-computing-can-bring-new-value/)
+ranging from simple [data collection](https://blog.equinix.com/blog/2017/11/29/how-iot-data-collection-and-aggregation-with-local-event-processing-work/)
+all the way to [Edge Analytics and AI](https://www.datanami.com/2019/02/04/exploring-artificial-intelligence-at-the-edge/).
+While EVE provides an incredible amount of functionality to support these use
+cases, it needs to be told exactly how to orchestrate these functional building blocks
+by a *Controller*. In other words, without Controller telling a fleet of Edge
+Nodes what to do, EVE by itself may be capable, but it practically useless.
+
+This is a fundamental point of EVE's design: an end user (lets call her Diane
+The DevOps) trying to implement an Edge Computing scenario never communicates
```
Comment (deitch) [inline docs/README.md]: "Diane the DevOps" - love it!

File: docs/README.md [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+intent to a controller using a Controller API and it is then the job of a
+controller to build an execution plan for a fleet of Edge Nodes running EVE.
+When each instance of EVE contacts the controller using EVE API it will receive
+its own portion of the overall execution plan (note that EVE API communications
+are  always initiated by EVE, never by the Controller):
+
+![EVE Controller](pics/EVEController.png)
+
+A good way to think about this is that Controller is to EVE what Kubernetes is
+to Docker Engine.
```
Comment (deitch) [inline docs/README.md]: This is a good analogy which many will grasp right away. I am a bit concerned that it makes EVE look like, "oh, someone just trying to do k8s at the edge" (which, to some degree, it is), but EVE is much more than that. I would leave it for now, but keep an eye on the impact.

File: docs/README.md [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+
+An API between EVE and its Controller is considered a [public API of Project EVE](../api).
+Refer to [API documentation](../api/API.md) for detailed information on API
+end points and message format.
+
+EVE design makes the trust between an EVE instance and its controller extremely
+important to the overall security of your Edge Nodes. An attack vector of
+impersonating an EVE controller or an EVE instance can be used for extremely
+malicious effects (like turning a [fleet of Edge Nodes into a botnet](https://www.cloudflare.com/learning/ddos/glossary/mirai-botnet/)).
+EVE takes great care of making sure that an EVE instance can alwats trust its
```
Comment (deitch) [inline docs/README.md]: "great care in making sure... can always trust" (type: "alwats")

File: docs/README.md [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+Controller and the Controller can always trust EVE. All of these techniques
+are documented in great details in the [EVE Security section](#security) of
```
Comment (deitch) [inline docs/README.md]: "document in detail in"

File: docs/README.md [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+this document.
+
+A decision of what controller to chose for maintaining a fleet of Edge Nodes
+running EVE currently boils down to two options:
+
+* Hyperscale, commercial [ZEDEDA Controller](https://zededa.com/demo-now/)
```
Comment (deitch) [inline docs/README.md]: I love ZedCloud, but this is the EVE docs, part of LFEdge which will, eventually, have more. I don't want to turn them off. Can we change this to:

* Hyperscale, commercial, provided by LFEdge members, currently only [ZEDEDA Controller](https://zededa.com/demo-now) (more to follow)

?

File: docs/README.md [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+interfaces. After all, EVE can only receive its configuration object from a
+Controller via network-based API call. This, in turn, requires at least a single
+functional hardware network interface and those can only be enabled after we
+receive configuration object from the Controller.
+
+EVE solves this problem by following a two-pronged approach:
+
+1. Having an out-of-band set of mechanisms that allow for the first
+   configuration object to be made available right after EVE instance boots
+2. Maintaining a sophisticated fallback semantics for the series of
```
Comment (deitch) [inline docs/README.md]: "Maintaining sophisticated" (drop "a")

File: docs/README.md [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+   configuration objects that EVE keeps receiving from the Controller
+
+The first part helps EVE bootstrap and configure itself during the initial boot
+sequence, the second part makes sure that EVE never ends up in a situation where
```
Comment (deitch) [inline docs/README.md]: "second part ensures"

File: docs/README.md [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+It must be noted that currently we are still not quite there with out-of-band
+mechanism for delivery of EVE's configuration object. While ideal EVE
+implementation would simply be able to consume exactly the same protobuf encoded
+binary blob that it receives from the Controller, currently we still have to
+rely on an ad-hoc collection of configuration files that serve the same purpose.
+We expect these configuration files to go away relatively quickly, but for now
+EVE is still stuck with at least *DevicePortConfig/global.json* and
+*GlobalConfig/global.json* and they are documented in [legacy configuration](CONFIG.md).
+See [the following FAQ entry](FAQ.md) for how to manage both of these
+legacy files.
```
Comment (deitch) [inline docs/README.md]: This is important. The fact that we have a separate file documenting it, and not just buried in here, is a big deal. We may even need to highlight it elsewhere, but nicely done.

File: docs/README.md [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+currently does *not* support network [PXE booting](https://en.wikipedia.org/wiki/Preboot_Execution_Environment)
+and given that Edge Nodes are typically deployed with very rudimentary networks
+attached to them it is very unlikely that EVE will support that type of
+booting. That said, it may be possible to use EVE's installer in a network boot
+scenario to install EVE on Edge Nodes when they first do a network boot in a lab
+or provisioning environment.
+
+During the installation phase, EVE assumes the target storage medium to be either
+empty or have been formatted with [GPT](https://en.wikipedia.org/wiki/GUID_Partition_Table)
+that EVE can recognize and extend. Either way, after succcesfull installation
```
Comment (deitch) [inline docs/README.md]: "successful".

File: docs/README.md [hunk 19]
```diff
@@ -0,0 +1,325 @@
...
+you will find EVE claiming the following entries int the GPT:
```
Comment (deitch) [inline docs/README.md]: "EVE will have created the following entries in the GPT"

File: docs/README.md [hunk 20]
```diff
@@ -0,0 +1,325 @@
...
+
+* EFI System (vfat formatted read-only filesystem that contains GRUB trampoline)
+* IMGA (first copy of EVE's read-only root filesystem)
+* IMGB (second copy of EVE's read-only root filesystem)
+* CONFIG (vfat formatted read-only filesystem that contains Edge Node Identity specific configuration)
+* P3 (read-write filesystem that extends to the rest of the available storage medium)
+
+An astute reader may notice that the above partitioning scheme is very reminiscent
```
Comment (deitch) [inline docs/README.md]: We probably can get rid of "astute reader". It works very well conversationally, less so in writing. Simpler as:

"The partitioning scheme is similar to how..."

File: docs/README.md [hunk 21]
```diff
@@ -0,0 +1,325 @@
...
+of how [CoreOS does its paritioning](https://coreos.com/os/docs/latest/sdk-disk-partitions.html).
+We are, indeed, leveraging the good work that CoreOS folks have done so much so
```
Comment (deitch) [inline docs/README.md]: "EVE does, indeed, leverage the good work..."

File: docs/README.md [hunk 22]
```diff
@@ -0,0 +1,325 @@
...
+that we're using their patches for the GRUB bootloader to provide rootfs upgrade
+functionality between IMGA and IMGB root filesystems. Unlike CoreOS, though, EVE
```
Comment (deitch) [inline docs/README.md]: Heh, I like this!

File: docs/README.md [hunk 23]
```diff
@@ -0,0 +1,325 @@
...
+firmware. Finally, it must be said that EVE doesn't support legacy partitioning
+schemes (such as MBR).
+
+EVE's root filesystem is hosted in both IMGA and IMGB partitions. This allows
+a running EVE instance to safely update itself. If you want to know more about
+how this process works read [baseimage update](BASEIMAGE-UPDATE.md) documentation.
+
+Both IMGA and IMGB host a filesystem containing the following software components
+
+* Second stage [GRUB bootloader](../pkg/grub) and [its configuration](../pkg/grub/rootfs.cfg) in /EFI folder
```
Comment (deitch) [inline docs/README.md]: Put /EFI in backticks please

File: docs/README.md [hunk 24]
```diff
@@ -0,0 +1,325 @@
...
+* [Hypervisor](../pkg/xen) and [Linux Kernel](../pkg/kernel) in /boot folder
```
Comment (deitch) [inline docs/README.md]: And for /boot please

File: docs/README.md [hunk 25]
```diff
@@ -0,0 +1,325 @@
...
+* [EVE microservices](../pkg/pillar) in /containers folder
```
Comment (deitch) [inline docs/README.md]: Ditto

File: docs/README.md [hunk 26]
```diff
@@ -0,0 +1,325 @@
...
+* Minimalistic [Alpine Linux](https://alpinelinux.org)-derived UNIX environment
+
+EVE microservices are ultimately responsible for providing the entire EVE
+experience and they are described in greater details below.
+
+## Runtime Lifecycle
+
+EVE boots as any traditional operating system and it requires UEFI firmware.
```
Comment (deitch) [inline docs/README.md]: "EVE boots as any traditional operating system that supports UEFI."

File: docs/README.md [hunk 27]
```diff
@@ -0,0 +1,325 @@
...
+kernel the rest of the runtime sequence relies solely on what happens with EVE
+microservices. All of these services are available under */containers/services*
+folder in IMGA or IMGB root filesystem. Their lifecycle is currently managed
+by [linuxkit init system](https://github.com/linuxkit/linuxkit/tree/master/pkg/init)
+and monitored by a built-in [software watchdog](../pkg/watchdog). Software
+watchdog, in turn, [relies](../pkg/pillar/scripts/device-steps.sh#L56) on
+Edge Node's hardware watchdog to cope with unexpected software failures.
+
+All EVE microservices are expected to communicate with each other by using
+our own [IPC mechanism](COMMS.md).
```
Comment (deitch) [inline docs/README.md]: "by using the EVE dedicated IPC mechanism"

File: docs/README.md (hunk 1 above)
Comment (rvs) [reply docs/README.md]: Thanks @deitch !

File: docs/README.md (hunk 2 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 3 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 4 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 5 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 6 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 7 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 8 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 9 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 10 above)
Comment (rvs) [reply docs/README.md]: ;-)

File: docs/README.md (hunk 11 above)
Comment (rvs) [reply docs/README.md]: good point, but lets leave it for now as you say

File: docs/README.md (hunk 12 above)
Comment (rvs) [reply docs/README.md]: 10x! This is super annoying since I didn't see it highlighted by my spell checker :-(

File: docs/README.md (hunk 13 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 14 above)
Comment (rvs) [reply docs/README.md]: great point. fixed.

File: docs/README.md (hunk 15 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 16 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 19 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 20 above)
Comment (rvs) [reply docs/README.md]: good point! ;-) fixed.

File: docs/README.md (hunk 21 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 22 above)
Comment (rvs) [reply docs/README.md]: 10x!

File: docs/README.md (hunk 23 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 24 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 25 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 26 above)
Comment (rvs) [reply docs/README.md]: fixed

File: docs/README.md (hunk 27 above)
Comment (rvs) [reply docs/README.md]: fixed

**Pull request merged**